source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py

```diff
@@ -0,0 +1,601 @@
+                params=self._cached_critic_network_params_detached,
+                target_params=self.target_critic_network_params,
+            )
+            advantage = tensordict.get(self.tensor_keys.advantage)
+        if self.normalize_advantage and advantage.numel() > 1:
+            loc = advantage.mean()
+            scale = advantage.std().clamp_min(1e-6)
+            advantage = (advantage - loc) / scale
+
+        log_weight, dist, kl_approx = self._log_weight(tensordict)
+        # ESS for logging
+        with torch.no_grad():
+            # In theory, ESS should be computed on particles sampled from the same source. Here we sample according
+            # to different, unrelated trajectories, which is not standard. Still it can give a idea of the dispersion
+            # of the weights.
+            lw = log_weight.squeeze()
+            ess = (2 * lw.logsumexp(0) - (2 * lw).logsumexp(0)).exp()
+            batch = log_weight.shape[0]
+
+        gain1 = log_weight.exp() * advantage
+
+        log_weight_clip = log_weight.clamp(*self._clip_bounds)
+        clip_fraction = (log_weight_clip != log_weight).to(log_weight.dtype).mean()
+        ratio = log_weight_clip.exp()
+        gain2 = ratio * advantage
+
+        gain = torch.stack([gain1, gain2], -1).min(dim=-1)[0]
+        td_out = TensorDict({"loss_objective": -gain.mean()}, batch_size=[])
+        td_out.set("clip_fraction", clip_fraction)
+
+        with self.actor_network_params.to_module(self.actor_network) if self.functional else contextlib.nullcontext():
+            current_dist = self.actor_network.get_dist(tensordict)
+            td_out.set("action_noise", current_dist.scale.mean())
+        try:
+            kl = torch.distributions.kl.kl_divergence(previous_dist, current_dist).mean()
+        except NotImplementedError:
+            x = previous_dist.sample(1)
+            kl = (previous_dist.log_prob(x) - current_dist.log_prob(x)).mean(0)
+        kl = kl.unsqueeze(-1)
+        td_out.set("kl", kl)
+        if self.entropy_bonus:
+            entropy = self.get_entropy_bonus(dist)
+            td_out.set("entropy", entropy.detach().mean())  # for logging
+            td_out.set("kl_approx", kl_approx.detach().mean())  # for logging
+            td_out.set("loss_entropy", -self.entropy_coef * entropy.sum(dim=-1).mean())
+        if self.critic_coef:
+            loss_critic, value_clip_fraction = self.loss_critic(tensordict)
+            td_out.set("loss_critic", loss_critic)
+            if value_clip_fraction is not None:
+                td_out.set("value_clip_fraction", value_clip_fraction)
+
+        td_out.set("ESS", _reduce(ess, self.reduction) / batch)
+        td_out = td_out.named_apply(
+            lambda name, value: (
+                _reduce(value, reduction=self.reduction).squeeze(-1) if name.startswith("loss_") else value
+            ),
+            batch_size=[],
+        )
+        return td_out
+
+
+class TrainerWrapper(Trainer):
```

Comment:
The trainer class hasn't been touched in a long time, as discussed I'm considering subclassing it for the various losses we have starting with PPO